src/Orleans/Providers/IEventStorageProvider.cs

```diff
@@ -0,0 +1,117 @@
+using System.Threading.Tasks;
+using Orleans.Providers;
+using Orleans.Runtime;
+using System.Collections.Generic;
+
+namespace Orleans.Storage
+{
+    /// <summary>
+    /// Interface to be implemented for a event storage provider able to read from and append to streams
+    /// representing the state of Journaled Grains
+    /// </summary>
+    public interface IEventStorageProvider : IProvider, IEventStorage
+    {
+        /// <summary>Logger used by this event storage provider instance.</summary>
+        /// <returns>Reference to the Logger object used by this provider.</returns>
+        /// <seealso cref="Logger"/>
+        Logger Log { get; }
+
+        /// <summary>
+        /// The provider can specify a default stream name to use for a given grain.
+        /// </summary>
+        /// <param name="grainType">The type of the grain</param>
+        /// <param name="grainReference">The grain reference for the grain</param>
+        /// <returns></returns>
+        string DefaultStreamName(Type grainType, GrainReference grainReference);
+    }
+
+    /// <summary>
+    /// The event store interface, as implemented by event storage providers
+    /// </summary>
+    /// 
+    public interface IEventStorage
+    {
+        /// <summary>
+        /// Returns a handle to an event stream that can be used to perform operations on it.
+        /// </summary>
+        /// <param name="streamName">The name of the stream</param>
+        /// <returns>A handle for performing operations on the stream</returns>
+        IEventStreamHandle GetEventStreamHandle(string streamName);
+    }
+
+    /// <summary>
+    /// The event stream interface, as implemented by event storage providers. 
+    /// Should be disposed when no longer needed.
+    /// </summary>
+    public interface IEventStreamHandle : IDisposable
+    {
+        /// <summary>
+        /// The name of this event stream.
+        /// </summary>
+        string StreamName { get; }
+
+        /// <summary>
+        /// Returns the number of events in the event stream, which corresponds to the version number.
+        /// Returns zero for a not-yet-created stream or for a deleted stream.
+        /// </summary>
+        /// <returns>a nonnegative integer representing the number of events in the stream.</returns>
+        Task<int> GetVersion();
+
+        /// <summary>
+        /// Loads a sequence of events from the event stream.
+        /// </summary>
+        /// <typeparam name="E">The base class for the events</typeparam>
+        /// <param name="startAtVersion">The version (stream position) at which to start</param>
+        /// <param name="endAtVersion">The version (stream position) at which to end, or null to return up and including the most recent</param>
+        /// <returns>A <see cref="EventStreamSegment{E}"/> structure containing a subrange of the event stream.</returns>
+        Task<EventStreamSegment<E>> Load<E>(int startAtVersion = 0, int? endAtVersion = null);
```

Comment:
And also since this is a generic method (but it's not a generic type, which I'm not suggesting to use a generic type either, I'd like to drop this open generic altogether), then what `E` will be used would be completely up to the infrastructure and user code who's calling this, which is not always straightforward when persisting and reading from storage. So the IEventStreamHandle shouldn't make any assumptions about how important is that implicit `E` typeparam passed in. If that type was important for the serialization logic, then it should be an explicit parameter.

All this just to reiterate that we should drop this and go back to `object` or `IEvent` or whatever, but not an open generic.